filesystem/index.go

```diff
@@ -0,0 +1,158 @@
+package fs
+
+import (
+	"context"
+	"io"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/ipfs/go-unixfs"
+)
+
+// This file represents a partial and theoretical
+// pkg-level implementation of a filesystem{}
+
+var (
+	pkgRoot FileSystem
+	closers []io.Closer
+)
+
+func init() {
+
+	// we depend on data from the coreapi to initalize our API nodes
+	// so fetch it or something and store it on the FS
+	daemon := fallbackApi()
+	ctx := deriveCtx(daemon.ctx) // if the daemon cancels, so should we
+
+	pkgRoot = NewFileSystem(ctx)
+	for _, pair := range [...]struct {
+		string
+		ParseFn
+	}{
+		{"/", rootParser},
+		{"/ipfs", pinRootParser},  // requests for "/ipfs" are directed at pinRootParser(ctx, requestString)
+		{"/ipfs/", coreAPIParser}, // all requests beneath "/ipfs/" are directed at coreAPIParser(ctx, requestString)
+		{"/ipns", keyRootParser},
+		{"/ipns/", nameAPIParser},
+		{filesRootPrefix, filesAPIParser},
+	} {
+		closer, err := pkgRoot.Register(pair.string, pair.ParseFn)
+		if err != nil {
+			if err == errRegistered {
+				panic(fmt.Sprtinf("%q is already registered", pair.string))
+			}
+			panic(fmt.Sprtinf("cannot register %q: %s", pair.string, err))
+		}
+
+		// store these somewhere important and call them before you exit
+		// this will release the namespace at the FS level
+		closers = append(closers, closer) // in our example we do nothing with them :^)
+	}
+}
+
+func NewDefaultFileSystem(parentCtx context.Context) (FileSystem, error) {
+	// something like this
+	fsCtx := deriveFrom(parentCtx)
+	// go onCancel(fsCtx) { callClosers() } ()
```

Comment:
I'd use goprocess instead of contexts since it allows for much nicer closing logic (and we already use it for quite a few things)